[ICDS Dashboard]: AP Data Pull

ShubhamBansal1997 opened the pull request:
- Github Issue:

<!--- Provide a link to the ticket or document which prompted this change -->

##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
https://app.asana.com/0/1169088777790552/1169099451366333/f
##### FEATURE FLAG
<!--- If this is specific to a feature flag, which one? -->

##### PRODUCT DESCRIPTION
<!--- For non-invisible changes, describe user-facing effects. -->

ShubhamBansal1997 committed:
[build]: ap data pull

- Github Issue:

ShubhamBansal1997 commented on custom/icds_reports/management/commands/sql_scripts/ap_data_pull_1.sql:
```diff
@@ -0,0 +1,28 @@
+    awc.supervisor_name,
+    awc.awc_name,
+    awc.awc_site_code,
+    CASE WHEN (SUM(CASE WHEN chm.recorded_weight IS NOT NULL THEN 1 ELSE 0 END)/count(*))::float>=0.6 THEN 'Y'
```
@Rohit25negi Is it good to do the calculation here, like is there any chance of divide by 0 case here, if that's how this code gonna handle?

ShubhamBansal1997 commented on custom/icds_reports/management/commands/sql_scripts/ap_data_pull_1.sql:
```diff
@@ -0,0 +1,28 @@
+    CASE WHEN (SUM(CASE WHEN chm.recorded_weight IS NOT NULL THEN 1 ELSE 0 END)/count(*))::float>=0.6 THEN 'Y'
+        ELSE 'N'
+    END as gm_%(month_execute)s,
+    CASE WHEN (SUM(CASE WHEN chm.num_rations_distributed>=21 THEN 1 ELSE 0 END)/SUM(CASE WHEN chm.age_tranche::int>=6 THEN 1 ELSE 0 END))>=0.6 THEN 'Y'
```
Same here

ShubhamBansal1997 commented on custom/icds_reports/management/commands/sql_scripts/ap_data_pull_2.sql:
```diff
@@ -0,0 +1,18 @@
+    awc.supervisor_name,
+    awc.awc_name,
+    awc.awc_site_code,
+    CASE WHEN (agg.vhnd_conducted IS NOT NULL AND agg.vhnd_conducted>0) THEN 'Y'
```
@Rohit25negi Is this correct?

Rohit25negi replied:
looks correct. Can you use `num_awcs_conducted_vhnd` column instead? I doubt vhnd_conducted goes that back in time.

Rohit25negi commented on custom/icds_reports/management/commands/sql_scripts/ap_data_pull_2.sql:
```diff
@@ -0,0 +1,18 @@
+    awc.awc_site_code,
+    CASE WHEN (agg.vhnd_conducted IS NOT NULL AND agg.vhnd_conducted>0) THEN 'Y'
+        ELSE 'N'
+    END as vhnd_conducted_month_executes
```
good to add which month in the column name as previous queries.

Rohit25negi commented on custom/icds_reports/management/commands/ap_data_pull.py:
```diff
@@ -0,0 +1,52 @@
+            path = os.path.join(os.path.dirname(__file__), 'sql_scripts',
+                                'ap_data_pull_{}.sql'.format(i))
+            context = {
+                'month_execute': date.strftime("%Y-%m-%d").replace("-", "_")
```
Why are you replacing `-` with `_`? Wont it fail because in the query it would become 
`where month='2019_04_01'` instead of `where month='2019-04-01'`

Rohit25negi commented on custom/icds_reports/management/commands/sql_scripts/ap_data_pull_1.sql:
```diff
@@ -0,0 +1,28 @@
+COPY(SELECT
```
For the simplification of the management command. i think you can combine all the SQL queries into a single SQL file. I was trying to split them into different files so that if we were able to write the simple one early, could  have been run at night or in the morning. Itself. But I think its better now to just merge different csv files.

Rohit25negi reviewed (changes requested):
Left a bunch of correction and optimization comments.

Rohit25negi commented:
@ShubhamBansal1997  Just want to put in. you know right THR is also given to ccs cases as well. So we would need to pull the thr info from ccs_record_monthly table as well similarly as you did for child_health_monthly.

ShubhamBansal1997 committed:
[fix]: minor fixes after review

- Github Issue:

ShubhamBansal1997 committed:
[refactor]:
- Shifted the montly queries to quaters
- Added logging
- Added the thr from the ccs record
- Added the launched and not launched status month wise
- Storing files for each quater in the month

- Github Issue:

Rohit25negi commented on custom/icds_reports/management/commands/sql_scripts/ap_data_pull_2.sql:
```diff
@@ -0,0 +1,83 @@
+COPY(SELECT
```
this query would needs to prepared like this. Else it would give multiple rows per awc


```
COPY(SELECT
    awc.state_name,
    awc.district_name,
    awc.block_name,
    awc.supervisor_name,
    awc.awc_name,
    awc.awc_site_code,
    SUM(CASE WHEN agg.num_awcs_conducted_vhnd>0 AND agg.month='2019-04-01' THEN 1 ELSE 0 END) as vhnd_conducted_April_2019,
    SUM(CASE WHEN agg.num_launched_awcs=1 AND agg.month='2019-04-01' THEN 1 ELSE 0 END) as     launched_status_April_2019,                                                                                       
    SUM(CASE WHEN agg.num_awcs_conducted_vhnd>0 AND agg.month='2019-05-01' THEN 1 ELSE 0 END) as vhnd_conducted_may_2019,
    SUM(CASE WHEN agg.num_launched_awcs=1 AND agg.month='2019-05-01' THEN 1 ELSE 0 END) as     launched_status_may_2019,  
    SUM(CASE WHEN agg.num_awcs_conducted_vhnd>0 AND agg.month='2019-06-01' THEN 1 ELSE 0 END) as vhnd_conducted_June_2019,
    SUM(CASE WHEN agg.num_launched_awcs=1 AND agg.month='2019-06-01' THEN 1 ELSE 0 END) as     launched_status_June_2019
FROM
    "awc_location_local" awc LEFT JOIN "agg_awc" agg ON (
        awc.supervisor_id=agg.supervisor_id and
        awc.doc_id = agg.awc_id AND
        awc.supervisor_id=agg.supervisor_id AND
        agg.month in ('2019-04-01','2019-05-01','2019-06-01')
    )
    where awc.state_id='f98e91aa003accb7b849a0f18ebd7039' and awc.aggregation_level=5
    group by awc.state_name,
    awc.district_name,
    awc.block_name,
    awc.supervisor_name,
    awc.awc_name,
    awc.awc_site_code
    
    
    ) TO '/tmp/ap_data_pull_2.csv' DELIMITER ',' CSV HEADER;

```


this 1/0 can be formated during consolidation

ShubhamBansal1997 replied:
updated

Rohit25negi commented on custom/icds_reports/management/commands/sql_scripts/ap_data_pull_1.sql:
```diff
@@ -0,0 +1,65 @@
+    awc.block_name,
+    awc.supervisor_name,
+    awc.awc_name,
+    awc.awc_site_code) TO '/tmp/%(name)s/ap_data_pull_1.csv' DELIMITER ',' CSV HEADER
```
will this give error if these folders are not already created?

ShubhamBansal1997 replied:
Not sure, let's try something basic on the server

Rohit25negi reviewed (changes requested):
left 3 minor changes. After those looks good to go

ShubhamBansal1997 committed:
[fix]: minor fix

- Github Issue:

ShubhamBansal1997 commented on custom/icds_reports/management/commands/sql_scripts/ap_data_pull_1.sql:
```diff
@@ -0,0 +1,65 @@
+    awc.supervisor_name,
+    awc.awc_name,
+    awc.awc_site_code,
+    CASE WHEN (SUM(CASE WHEN chm.nutrition_status_weighed=1 AND chm.age_tranche::int<=36 AND month='%(month_1)s' THEN 1 ELSE 0 END)/SUM(CASE WHEN chm.wer_eligible=1 AND chm.age_tranche::int<=36 AND month='%(month_1)s' THEN 1 ELSE 0 END))::float>=0.6 THEN 'Y'
```
updated

ShubhamBansal1997 commented on custom/icds_reports/management/commands/ap_data_pull.py:
```diff
@@ -0,0 +1,79 @@
+            },
+            {
+                "name": "january_2020_march_2020",
+                "column_1": "october_2019",
```
updated

ShubhamBansal1997 committed:
[fix]: minor typo fix

- Github Issue:

Rohit25negi reviewed (approved):
looks good. Good to go. once the folder thing is sorted.

ShubhamBansal1997 committed:
[fix]: added the script used to consolidate the data

- Github Issue:

ShubhamBansal1997 committed:
[build]: added the script used for the consolidation

- Github Issue:

Authored-by: Shubham Bansal <shubhambansal17@hotmail.com>
Signed-off-by: Shubham Bansal <shubhambansal17@hotmail.com>
